Add phase entities to Enphase Envoy

catsmanac opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Enphase Envoy metered with current transformer (CT) clamps installed on multiple phases provides data at the phase level in addition to the overall summary. This PR add entities for phase production and consumption data. Entity names are the same as the overall summary entities with phase names l1, l2 or l3 post-fixed.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30818

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

catsmanac committed:
add phase entities to Enphase Envoy

catsmanac committed:
Implement review feedback for translation strings

joostlek commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -411,12 +503,38 @@ def __init__(
+        ct_consumptionmeter = coordinator.envoy.consumption_meter_type
+        phase_count = coordinator.envoy.phase_count
+        # Display phase count and found CT meters in model
+        model = (
+            "Envoy "
+            # Show CT Phase mode if CT is found
+            + (
+                f", phase mode: {coordinator.envoy.phase_mode}"
+                if ct_consumptionmeter is not None
+                else ""
+            )
+            # show number of phases found in Model description
+            + f" , phases: {str(phase_count)}"
+            # Show production CT found if 1 CT and no consumption ct of if more CT found
+            + (
+                ", production CT"
+                if (ct_count == 1 and ct_consumptionmeter is None) or (ct_count > 1)
+                else ""
+            )
+            # Add type of consumption CT to model if one was found
+            + (f", {ct_consumptionmeter} CT" if ct_consumptionmeter is not None else "")
+            + (", no CT" if ct_count == 0 else "")
+        )
```
I have mixed feelings about this part. It's very unclear what is happening when. I am leaning towards extracting this into a function, but that would require a lot of parameters. Or you would have to pass in the coordinator. 

(The biggest factor that makes this worse are the inline ternaries)

catsmanac replied:
Intent is to show phase features with the model. Maybe this makes it more clear:
```Py
    model = f"Envoy, phases: {phase_count}"
    if ct_consumptionmeter:
        # Add CT phase mode to model if consumption CT is found
        model = f"{model}, phase mode: {coordinator.envoy.phase_mode}"
        # Add type of consumption CT to model if CT is found
        model = f"{model}, {ct_consumptionmeter} CT"
    if (ct_count == 1 and ct_consumptionmeter is None) or (ct_count > 1)
        # Add Production CT to model if present
        # if only 1 ct and no consumption ct found or more as 1 ct found
        model = f"{model}, production CT"
    if ct_count == 0:
        model = f"{model}, no CT"
```

joostlek replied:
I think it looks better, but I do think we should put it in a separate function

catsmanac replied:
moved to seperate function

joostlek commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -411,12 +503,38 @@ def __init__(
+            model=model,
             name=coordinator.name,
             sw_version=str(coordinator.envoy.firmware),
+            hw_version=coordinator.envoy.part_number or "Unknown",
```
Can part_number be None? In that case omit the `or "Unknown"`, this will just hide it in the UI (Like, "Unknown" doesn't add value, so might as well just hide it)

catsmanac replied:
That section came from original code, only change is to show it in hardware version rather then model and change Envoy to Unknown.

```py
            model=coordinator.envoy.part_number or "Envoy",
```

```py
            hw_version=coordinator.envoy.part_number or "Unknown",
```
So it probably made sense in old setup and can be removed here.

catsmanac replied:
Unknown removed

joostlek commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -446,6 +564,40 @@ def native_value(self) -> int | None:
+    @property
+    def native_value(self) -> int | None:
+        """Return the state of the sensor."""
+        if self.data.system_production_phases is not None:
```
Can you maybe elaborate a bit more about why we do this? Is this a way to remove an assert (which we have too many of, I agree)?

Or does this mean we only show unknown at devices which don't support this? When can this be None?

(this question also applies for the Consumption one)

(if this is intended, I think inverting it would be better)

```suggestion
        if self.data.system_production_phases is None:
            return None
        ...
```

catsmanac replied:
Yes they can be none and will be none for Envoy without CT or non-metered versions. I'll apply your reverted proposal and do the same for the one below to get rid of the assert
```py
            if system_production is None:
                return None
            return self.entity_description.value_fn(system_production)
```

joostlek replied:
But if someone has no CT, shouldn't we omit these entities from being added in the first place? This way they aren't fed up with X amount of entities that are Unknown

catsmanac replied:
Yes that is the intent. If none is returned no entity is created.

catsmanac replied:
refactored it

joostlek commented on tests/components/enphase_envoy/test_config_flow.py:
```diff
@@ -361,3 +363,55 @@ async def test_reauth(hass: HomeAssistant, config_entry, setup_enphase_envoy) ->
+    )
+    assert result2["result"].unique_id == "1234"
+
+    envoy = hass.data[DOMAIN][result2["result"].entry_id].envoy
```
We should not access internals during tests.

I am not up to date on the test state of the integration, but these ideally would be sensor tests as in 
1. We set up the integration with this as source data
2. Observe the created states and check if they are what we expect
3. profit

catsmanac replied:
I'll change to use   state = hass.states.get("sensor.whatever")

catsmanac replied:
May require more overhaul of current test. Currently the device is set up, but no entities, so no states. The flexit_bacnet example does this, but is setup different as enphase currently.

bdraco replied:
While it would be nice to have these tests, its better for them to be done in another PR where we create a `test_sensor.py`

catsmanac replied:
Ok, I'll revert the test change. With the new CT data in the 1.18 library version there will be another HA PR for net-consumption and net-production entities to use with Energy dashboard that need sensor testing as well. So a good foundation for that would be neat.

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -137,6 +138,42 @@ class EnvoyProductionSensorEntityDescription(
+) -> list[EnvoyProductionPhaseSensorEntityDescription]:
+    """Build phase sensors from production sensors."""
+    return [
+        EnvoyProductionPhaseSensorEntityDescription(
+            key=f"{sensor.key}_l{phase + 1}",
+            translation_key=f"{sensor.translation_key}_phase",
+            native_unit_of_measurement=sensor.native_unit_of_measurement,
+            state_class=sensor.state_class,
+            device_class=sensor.device_class,
+            suggested_unit_of_measurement=sensor.suggested_unit_of_measurement,
+            suggested_display_precision=sensor.suggested_display_precision,
+            value_fn=sensor.value_fn,
+            on_phase=PhaseNames(PHASENAMES[phase]),
+            translation_placeholders={"phase_name": f"l{phase + 1}"},
+        )
```
I think `dataclasses.replace` could be used here instead

https://docs.python.org/3/library/dataclasses.html#dataclasses.replace

catsmanac replied:
It has an additional item `on_phase` not present in dataclass its getting the data from. Would that work?

bdraco replied:
sadly no. it looks like we are stuck with something complex

bdraco replied:
You might be able to do `dataclasses.asdict` and unpack with `**`

untested... but maybe this will work,

```python
        EnvoyProductionPhaseSensorEntityDescription(
            **dataclasses.asdict(sensor),
            key=f"{sensor.key}_l{phase + 1}",
            translation_key=f"{sensor.translation_key}_phase",
            on_phase=PhaseNames(PHASENAMES[phase]),
            translation_placeholders={"phase_name": f"l{phase + 1}"},
        )
```

catsmanac replied:
Doesn't fly, complains about key and translation_key having double values as they are in the asdict list as well. To simplify this, adding the on_phase field to the base class (and not using it) would allow for the dataclasses.replace use.

bdraco replied:
> adding the on_phase field to the base class (and not using it) would allow for the dataclasses.replace use.

That seems like a better idea to me. (typing it to `None` when not used)

catsmanac replied:
Done

catsmanac committed:
Enphase Envoy multiphase review changes

Move device name logic to separate function.
Refactor native value for phases
Use dataclasses.replace for phase entities, add on-phase to base class as well, no need for phase entity descriptions anymore

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -361,6 +416,25 @@ async def async_setup_entry(
         )
+    # For each production phase reported add production entities
+    if envoy_data.system_production_phases:
+        for usephase in envoy_data.system_production_phases:
```
Please separate words with _

catsmanac replied:
Done (need to get rid of this old habbit)

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -414,11 +488,46 @@ def __init__(
+            serial_number=self.envoy_serial_num,
         )
 
+    def envoy_model(self) -> str:
```
Can we move this into the library?  It seems like HA shouldn't need to know how to generate this

catsmanac replied:
Yes. What's best path forward, leave as is for now until we added it to the library and replace when library is bumped. Or just name it 'Envoy' as it was and later apply the change when library is updated.

bdraco replied:
It would be better to update the library, bump it in HA in another PR, and than rebase this PR

catsmanac replied:
Ok, I'll move it to the library first. My rebase track record is awfull with last 3 attempts spoiled so I'm not sure you should trust me with that action.

bdraco replied:
No need to actually re-base, it's usually better just to merge commit in, or hit the update branch button. re-base and fourth push pattern is not desirable because it makes reviews harder, and since we're going to squash anyway there is no benefit to doing it

catsmanac replied:
Removed it and for now made fail safe with hasattr test and if not use 'Envoy'

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -446,6 +555,46 @@ def native_value(self) -> int | None:
+        """Return the state of the sensor."""
+        if self.entity_description.on_phase is None:
+            return None
+        if self.data.system_production_phases is None:
```
Could combine these ifs since they do the same return

catsmanac replied:
Done

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -446,6 +555,46 @@ def native_value(self) -> int | None:
+        """Return the state of the sensor."""
+        if self.entity_description.on_phase is None:
+            return None
+        if self.data.system_consumption_phases is None:
```
As above

catsmanac replied:
Done

catsmanac committed:
Enphase Envoy reviewe feedback

Move model determination to library.
Revert states test for future split to sensor test.

bdraco committed:
Merge branch 'dev' into EnphaseEnvoyPhases-2

catsmanac committed:
Enphase_Envoy use model description from pyenphase library

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -361,6 +416,25 @@ async def async_setup_entry(
         )
+    # For each production phase reported add production entities
+    if envoy_data.system_production_phases:
+        for use_phase in envoy_data.system_production_phases:
```
```suggestion
        for use_phase, phase in envoy_data.system_production_phases.items():
```

bdraco replied:
```python
        entities.extend(
            EnvoyProductionPhaseEntity(coordinator, description)
            for description in PRODUCTION_PHASE_SENSORS
            if description.on_phase == use_phase
            for use_phase, phase in envoy_data.system_production_phases.items()
            if phase is not None
        )
```

I think this works as well

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -361,6 +416,25 @@ async def async_setup_entry(
+                )
+    # For each consumption phase reported add consumption entities
+    if envoy_data.system_consumption_phases:
+        for use_phase in envoy_data.system_consumption_phases:
+            if envoy_data.system_consumption_phases[use_phase] is not None:
+                entities.extend(
+                    EnvoyConsumptionPhaseEntity(coordinator, description)
+                    for description in CONSUMPTION_PHASE_SENSORS
+                    if description.on_phase == use_phase
+                )
```
```suggestion
        entities.extend(
            EnvoyConsumptionPhaseEntity(coordinator, description)
            for description in CONSUMPTION_PHASE_SENSORS
            if description.on_phase == use_phase
            for use_phase, phase in envoy_data.system_consumption_phases.items()
            if phase is not None
        )
```

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -361,6 +416,25 @@ async def async_setup_entry(
         )
+    # For each production phase reported add production entities
+    if envoy_data.system_production_phases:
+        for use_phase in envoy_data.system_production_phases:
+            if envoy_data.system_production_phases[use_phase] is not None:
+                entities.extend(
+                    EnvoyProductionPhaseEntity(coordinator, description)
+                    for description in PRODUCTION_PHASE_SENSORS
+                    if description.on_phase == use_phase
+                )
```
```suggestion
        entities.extend(
            EnvoyProductionPhaseEntity(coordinator, description)
            for description in PRODUCTION_PHASE_SENSORS
            if description.on_phase == use_phase
            for use_phase, phase in envoy_data.system_production_phases.items()
            if phase is not None
        )
```

bdraco replied:
But we can avoid iterating `PRODUCTION_PHASE_SENSORS` if we change it to a dict https://github.com/home-assistant/core/pull/108725/files#r1468660061

catsmanac replied:
Using PRODUCTION_PHASE_SENSORS as dict (above) condensed this to:
```py
        entities.extend(
            EnvoyProductionPhaseEntity(coordinator, description)
            for use_phase, phase in envoy_data.system_production_phases.items()
            for description in PRODUCTION_PHASE_SENSORS[use_phase]
            if phase is not None
        )
```

catsmanac replied:
tested but not committed yet. If ok I'll commit.

bdraco replied:
Looks much better. Go for it

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -133,15 +138,39 @@ class EnvoyProductionSensorEntityDescription(
+    ]
+
+
+PRODUCTION_PHASE_SENSORS = tuple(
```
It would be better to store this in a `dict` by `phase` so we can simplify below

catsmanac replied:
```py
    return {
        (PhaseNames(PHASENAMES[phase])): [
            replace(
                sensor,
                key=f"{sensor.key}_l{phase + 1}",
                translation_key=f"{sensor.translation_key}_phase",
                on_phase=PhaseNames(PHASENAMES[phase]),
                translation_placeholders={"phase_name": f"l{phase + 1}"},
            )
            for sensor in sensors
        ]
        for phase in range(0, 3)
    }
    ```

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -446,6 +522,48 @@ def native_value(self) -> int | None:
+    @property
+    def native_value(self) -> int | None:
+        """Return the state of the sensor."""
+        if (
+            self.entity_description.on_phase is None
+            or self.data.system_production_phases is None
+        ):
+            return None
```
Is this check needed here? I think we avoid creating the entity (or should) if its going to return here

catsmanac replied:
No check is not needed indeed, only for the Linters not liking the None part and avoiding assert.

catsmanac replied:
used
```py
        if TYPE_CHECKING:
            assert self.entity_description.on_phase
            assert self.data.system_production_phases
```

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -446,6 +522,48 @@ def native_value(self) -> int | None:
+    @property
+    def native_value(self) -> int | None:
+        """Return the state of the sensor."""
+        if (
+            self.entity_description.on_phase is None
+            or self.data.system_consumption_phases is None
+        ):
+            return None
```
Is this check needed here? I think we avoid creating the entity (or should) if its going to return here

catsmanac replied:
for linters,this or assert

catsmanac replied:
Changed to:
```py
        if TYPE_CHECKING:
            assert self.entity_description.on_phase
            assert self.data.system_production_phases
```

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -446,6 +522,48 @@ def native_value(self) -> int | None:
+            or self.data.system_consumption_phases is None
+        ):
+            return None
+        system_consumption = self.data.system_consumption_phases[
+            self.entity_description.on_phase
+        ]
+        if system_consumption is None:
```
Its already pretty long but I still think you save a line by using a walrus here

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -446,6 +522,48 @@ def native_value(self) -> int | None:
+            or self.data.system_production_phases is None
+        ):
+            return None
+        system_production = self.data.system_production_phases[
+            self.entity_description.on_phase
+        ]
+        if system_production is None:
+            return None
```
Its already pretty long but I still think you save a line by using a walrus here

catsmanac replied:
done

catsmanac committed:
Enphase_Envoy refactor Phase Sensors

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -133,15 +139,39 @@ class EnvoyProductionSensorEntityDescription(
+    }
+
+
+PRODUCTION_PHASE_SENSORS = phase_sensor_from_production_sensor(list(PRODUCTION_SENSORS))
```
I think we can drop the function here and build the constant directly

catsmanac replied:
This works:
```py
PRODUCTION_PHASE_SENSORS = {
    (PhaseNames(PHASENAMES[phase])): [
        replace(
            sensor,
            key=f"{sensor.key}_l{phase + 1}",
            translation_key=f"{sensor.translation_key}_phase",
            on_phase=PhaseNames(PHASENAMES[phase]),
            translation_placeholders={"phase_name": f"l{phase + 1}"},
        )
        for sensor in list(PRODUCTION_SENSORS)
    ]
    for phase in range(0, 3)
}
```

bdraco replied:
I think you could write it as
```python
PRODUCTION_PHASE_SENSORS = {
    (on_phase := PhaseNames(PHASENAMES[phase])): [
        replace(
            sensor,
            key=f"{sensor.key}_l{phase + 1}",
            translation_key=f"{sensor.translation_key}_phase",
            on_phase=on_phase,
            translation_placeholders={"phase_name": f"l{phase + 1}"},
        )
        for sensor in PRODUCTION_SENSORS
    ]
    for phase in range(0, 3)
}
```

catsmanac replied:
yep, works fine as well.

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -190,10 +223,35 @@ class EnvoyConsumptionSensorEntityDescription(
+    }
+
+
+CONSUMPTION_PHASE_SENSORS = phase_sensor_from_consumption_sensor(
```
I think we can drop the function here and build the constant directly

catsmanac committed:
Enphase_Envoy use walrus in phase sensor

bdraco commented:
I think its good to go from my perspective. I'll test it shortly

bdraco commented:
Works fine on my systems but they are all 2 phase

bdraco commented:
I approved it but didn't merge to give joostlek a chance top followup on his comment https://github.com/home-assistant/core/pull/108725#discussion_r1463424770 in case any more changes are needed

catsmanac commented:
> Works fine on my systems but they are all 2 phase

fwiw I ran it with my envoy simulator using the 3 phase fixture files from pyenphase.

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -361,6 +403,23 @@ async def async_setup_entry(
+        entities.extend(
+            EnvoyConsumptionPhaseEntity(coordinator, description)
+            for use_phase, phase in envoy_data.system_consumption_phases.items()
+            for description in CONSUMPTION_PHASE_SENSORS[PhaseNames(use_phase)]
```
I think we could simplify this further and make `use_phase` the key to it becomes `CONSUMPTION_PHASE_SENSORS[use_phase]` and the dict construction above is adjusted accordingly

Sorry for missing this one the first go.

catsmanac replied:
The linter didn't like the  
```py
for description in CONSUMPTION_PHASE_SENSORS[use_phase]
```
when the walrus was added and the `PhaseNames` had to be added. To optimize it we need to change the definition in the EnvoyConsumptionRequiredKeysMixin to str

```py
    on_phase: str | None
```
and make sure the key is str
```py
CONSUMPTION_PHASE_SENSORS = {
    (on_phase := str(PhaseNames(PHASENAMES[phase]))): [
```

bdraco replied:
No worries. Let's leave it as is
